File: GSS1/main.cpp
#include <iostream>
#include <vector>
using namespace std;

struct node {
  int v;
  node *l, *r;
  int li, ri;
};

vector<int> a;
node* build(int i, int j) {
  node* n = new node();
  n->v = 0;
  n->li = i;
  n->ri = j;
  n->v = a[i];
  if(j-i > 0) {
    int mid = (i+j)/2;
    n->l = build(i, mid);
    n->r = build(mid+1, j);
    n->v = max(n->l->v+n->r->v, max(n->l->v, n->r->v));
  }
  return n;
}

int solve(int x, int y, node* n) {
  if(y<n->li || x>n->ri) {
    return -1;
  }
  if(n->li >= x && n->ri <= y) {
    return n->v;
  }
  return max(solve(x,y,n->l), solve(x,y,n->r));
}

int main() {
  int n;
  cin >> n;
  a.resize(n+1);
  a[0]=0;
  for(int i=0;i<n;i++) {
    cin >> a[i+1];
  }
  node* tree = build(0, n);
  int numQueries;
  cin >> numQueries;
  for(int i=0;i<numQueries;i++) {
    int x, y;
    cin >> x >> y;
    cout << solve(x, y, tree) << endl;
  }
  return 0;
}

File: GSS1/main.go
package main

import "fmt"

type node struct {
	v      int
	l, r   *node
	li, ri int
}

func build(a []int, i int, j int) *node {
	n := node{}
	n.li = i
	n.ri = j
	for m := i; m <= j; m++ {
		if a[m] > n.v {
			n.v = a[m]
		}
	}
	if j-i > 0 {
		mid := (i + j) / 2
		n.l = build(a, i, mid)
		n.r = build(a, mid+1, j)
	}
	return &n
}

func solve(x, y int, n *node) int {
	if y < n.li || x > n.ri {
		return -1
	}
	if n.li >= x && n.ri <= y {
		return n.v
	}
	a := solve(x, y, n.l)
	b := solve(x, y, n.r)
	if a > b {
		return a
	}
	return b
}

func main() {
	var n int
	fmt.Scanln(&n)
	a := make([]int, n+1)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i+1])
	}
	tree := build(a, 0, n)
	var numQueries int
	fmt.Scan(&numQueries)
	for i := 0; i < numQueries; i++ {
		var x, y int
		fmt.Scanln(&x, &y)
		res := solve(x, y, tree)
		fmt.Println(res)
	}
}
